dali/c_api_2/data_objects.cc

```diff
@@ -0,0 +1,362 @@
+// Copyright (c) 2025, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "dali/c_api_2/data_objects.h"
+#include "dali/c_api_2/error_handling.h"
+
+namespace dali::c_api {
+
+RefCountedPtr<ITensor> ITensor::Create(daliBufferPlacement_t placement) {
+  Validate(placement);
+  switch (placement.device_type) {
+    case DALI_STORAGE_CPU:
+    {
+      auto tl = std::make_shared<Tensor<CPUBackend>>();
+      tl->set_pinned(placement.pinned);
+      if (placement.pinned)
+        tl->set_device_id(placement.device_id);
+      return Wrap(std::move(tl));
+    }
+    case DALI_STORAGE_GPU:
+    {
+      auto tl = std::make_shared<Tensor<GPUBackend>>();
+      tl->set_pinned(placement.pinned);
+      tl->set_device_id(placement.device_id);
+      return Wrap(std::move(tl));
+    }
+    default:
+      assert(!"Unreachable code");
+      return {};
+  }
+}
+
+RefCountedPtr<ITensorList> ITensorList::Create(daliBufferPlacement_t placement) {
+  Validate(placement);
+  switch (placement.device_type) {
+    case DALI_STORAGE_CPU:
+    {
+      auto tl = std::make_shared<TensorList<CPUBackend>>();
+      tl->set_pinned(placement.pinned);
+      if (placement.pinned)
+        tl->set_device_id(placement.device_id);
+      return Wrap(std::move(tl));
+    }
+    case DALI_STORAGE_GPU:
+    {
+      auto tl = std::make_shared<TensorList<GPUBackend>>();
+      tl->set_pinned(placement.pinned);
+      tl->set_device_id(placement.device_id);
+      return Wrap(std::move(tl));
+    }
+    default:
+      assert(!"Unreachable code");
+      return {};
+  }
+}
+
+ITensor *ToPointer(daliTensor_h handle) {
+  if (!handle)
+    throw NullHandle("Tensor");
+  return static_cast<ITensor *>(handle);
+}
+
+ITensorList *ToPointer(daliTensorList_h handle) {
+  if (!handle)
+    throw NullHandle("TensorList");
+  return static_cast<ITensorList *>(handle);
+}
+
+}  // namespace dali::c_api
+
+using namespace dali::c_api;  // NOLINT
+
+template <typename T>
+std::optional<T> ToOptional(const T *nullable) {
+  if (nullable == nullptr)
+    return std::nullopt;
+  else
+    return *nullable;
+}
+
+//////////////////////////////////////////////////////////////////////////////
+// Tensor
```

Comment:
It's missing impl for daliTensorGetShape, right?

Reply:
It was missing; added.

Reply:
Done.

Reply:
@mzient also missing:
daliTensorGetBufferPlacement
daliTensorListGetBufferPlacement
and source methods
daliTensorGetSourceInfo
daliTensorListGetSourceInfo